### ISSUE TITLE
DicomFile/DicomDataset should implement IDisposable interface

### ISSUE THREAD

**Issue by jackliums** [fo-dicom/fo-dicom#1117]
**Describe the bug**
DicomFile/DicomDataset should implement IDisposable interface since it could hold reference to the stream through StreamByteSource.

When DicomFile instance is created through Open/OpenAsync using a stream, it creates an instance of StreamByteSource with the originating stream and FileReadOptions.ReadLargeOnDemand by default. If the requested buffer is larger than the threshold requested, it creates StreamByteBuffer, which delays the reading of the byte until specifically requested.

Because DicomFile does not implement IDisposable even though it hold reference to the underlying stream, it makes the lifetime management of the stream more challenging since the caller has to keep track of both the stream and the DicomFile and only dispose the stream when it's done with the file itself.

**Environment**
Fellow Oak DICOM version: 4.0.6
OS: Windows 10 x64
Platform: .NET Core 3.1.402

**Comment by amoerie** [comment]
I agree, but it depends. When creating a DicomFile with a simple path, it keeps a FileReference in case it needs the FileStream again later to read the pixel data.
When creating a DicomFile with a stream, you are responsible for disposing of that stream after you've done everything you need to with the DicomFile. 

In any case, I do agree that it would make sense to make DicomFile IDisposable, and I think this would be a good change to consider for the v5 release, wouldn't you agree @gofal ?

For reference, someone had an actual issue about memory usage related to DicomFile disposing not too long ago: https://github.com/fo-dicom/fo-dicom/issues/982

**Comment by gofal** [comment]
Sure, I agree. This will be a great enhancement in v5.

**Comment by sjrulandch** [comment]
Any update on this? 
v5.1.3 seems to be the latest.

**Comment by gofal** [comment]
Not yet. We had a try, to make the classes in fo-dicom, that have ownership of streams,  to implement IDisposable. But this soon became massively spreading and nearly all of the classes contained in fo-dicom would have been IDisposable.
But propably it would work better if IDisposable would not be implemented from bottom up, but from up to bottom.
